### ISSUE TITLE
Tasks 16 and 19 do not converge

### ISSUE THREAD

**Issue by jimfleming** [jimfleming/recurrent-entity-networks#5]
Task 16 (basic induction) and task 19 (path-finding) overfit and the test loss did not converge after 10 runs with different random seeds:

<img width="879" alt="screen shot 2016-11-29 at 9 13 27 am" src="https://cloud.githubusercontent.com/assets/226553/20720473/204d24a2-b614-11e6-8f53-faf88dc46914.png">

Possibilities:

- There is a bug in the model.
- There is a bug in the data.
- There is a detail not mentioned in the paper.

If someone finds the discrepancy please send a PR or let me know here.

**Comment by lhoang29** [comment]
This may be unrelated but in my experience with memory networks, non-linear layers cause problems on these specific tasks. If you haven't tried it already, I'd play with selectively turning them off, i.e. those in eq (2) and (3) from the paper.

**Comment by jimfleming** [comment]
Yah, they do this in the children's book tasks. I'd prefer to reproduce the experiments in the paper but your suggestion is something worth trying, anyway.

EDIT: Note that the PReLUs are initialized to identity so those non-linearities are not present at the beginning of training.

**Comment by jimfleming** [comment]
Mikael got back to me. Two details I don't think I'm accounting for:

> We used separate masking weights for encoding the question and the sentences
> The PReLUs weights are all tied together, both across timesteps in the memory update and the output module.

They also open-sourced their implementation (https://github.com/facebook/MemNN/tree/master/EntNet-babi) so I'll go through and make sure the details are correct here.

EDIT:
Okay, the masking weights were already separated.

At a quick glance, the PReLU weights are _not_ tied in the memory or output modules: https://github.com/jimfleming/recurrent-entity-networks/blob/master/entity_networks/activations.py#L12

I presume the author either means there is a single shared alpha parameter (with a width of the layers) across the entire model or at least within a single module (memory and output).

**Comment by botev** [comment]
Have anyone tried to understand their code if there is anything different? I've reimplemented this in Theano (independently without looking at the code here) and I get similar issues on task 16 and 19 and generally did not manage to get on some tasks as good results as the paper.

**Comment by jimfleming** [comment]
I'm updating the code to work with TF 1.1. I've also modified the PReLU activations so they're tied. Will go through the relevant code and see if any details are missed and re-run these tasks.

**Comment by jimfleming** [comment]
Found a few more details:

- It appears the position encoding has a weight per token and per unit of the embedding, not just per token as I have it currently.
- Their normalization is slightly different: rather than taking the max between the norm and epsilon, they simply add 1 to values less than or equal to zero.

**Comment by botev** [comment]
@jimfleming are you looking at their torch code or and which normalisation are you referring to?

**Comment by jimfleming** [comment]
Yes, the normalization from their code. It's not in the paper since it's
for numerical stability.

**Comment by jimfleming** [comment]
Two more details from the code, not in the paper:

- Bias on U (no bias on V, W).
- Memory and keys initialized with embeddings from out-of-vocabulary words, rather that simply using separate vars for the keys. This should be equivalent but I updated it anyway. This is technically in the paper under "Training Details" but my interpretation was different:

> The initial values of the memory slots were initialized to the key values, which we found to help performance.

Still not converging, though, so my suspicion is that the data format is incorrect for those tasks.

**Comment by botev** [comment]
What do you mean that the "data format" is incorrect?
What we saw after running our own code and running their own code as well is that these tasks converge maybe 10-20% of the runs, all other runs fail miserably, which explains very well why in the paper they present the results for "only the best run".

**Comment by jimfleming** [comment]
Indeed, I've been running 10 successive trials per the paper. If you saw the failure rate of those specific tasks at 10-20% I may just have had poor seeds, especially given the TF code reproduces the other 18 of 20 tasks. I'll try another 10 runs but if it's just due to the seed I may diverge from the paper and adjust the initialization instead (Gaussian distribution is not ideal).

(By data format I meant there may be an error in prep_datasets.py.)

EDIT: Another 10 runs did not converge. A very quick spot check of the data appeared correct. Will try changing the initialization / lr scheme.

**Comment by jimfleming** [comment]
Alright, had some time to get back to this. Tasks 16 and 19 both converge readily using a cyclic learning rate with a min of 2e-4, max of 1e-2, and step size of 10 epochs. Also, update coming soon which will include cleaned code, upgrade to TF 1.1.0, and code compatible with training/prediction on Cloud ML Engine.

(@mikalyoung also found that lowering the learning rate to 1e-3 works well for Task 16. I was able to reproduce this on Task 16 but no luck with Task 19 until CLR.)

**Comment by botev** [comment]
Have you tried comparing against their Torch code, in terms of % of runs solving the task and not overfitting it? I would find it highly suspicious if you get it to work 100% of the runs, while their code does not. Additionally, could you point out to any research material for a motivation on CLR as I heard about that for the first time.

**Comment by jimfleming** [comment]
Eventually, I'll add best + std but my focus for this week was to get it to converge reliably for the two remaining tasks and updating to TF 1.1.

CLR comes from https://arxiv.org/abs/1506.01186 and I've been using it lately because it can help convergence. Previously I've found "SGD with warm restarts" to help with overfitting on small datasets but it requires more hyperparameters: https://arxiv.org/abs/1608.03983

> I would find it highly suspicious if you get it to work 100% of the runs, while their code does not.

I don't know yet if it works 100% but definitely out of a few consecutive runs. Skepticism is good but since simply lowering the learning rate improved convergence (this is easily testable) it shouldn't be surprising. Since we've diverged from the original model spec we're no longer comparing apples to apples, just results.

**Comment by botev** [comment]
> I don't know yet if it works 100% but definitely out of a few consecutive runs. Skepticism is good but since simply lowering the learning rate improved convergence (this is easily testable) it shouldn't be surprising. Since we've diverged from the original model spec we're no longer comparing apples to apples, just results.

It depends, what you mean by convergence. Their code always "converges"/solves on the training data. However, it overfits and you get really terrible valid/test error (50%). Lowering the learning rate indeed can stabilize and improve convergence, but magically translating to significantly better generalization and avoiding overfitting - I think not.

Here is a direct snippet from 10 runs from their Torch code results (note that I intentionally show the file names as it can tell you exactly what parameters I used including the seed):
```
$ cat task\=16-seed\=12589-model\=entnet-embed\=icmul+bow-edim\=100-initstd\=0.1-sdt_decay\=25-optim\=adam-sdt\=0.01-nhop\=1-tied\=0-T\=70-memslots\=20.log 
Sat Mar 18 16:01:53 2017:run 1 | train error = 0.0000 | valid error = 0.0101 | test error = 0.0071
Sat Mar 18 19:16:36 2017:run 2 | train error = 0.0396 | valid error = 0.5433 | test error = 0.5020
Sat Mar 18 22:31:11 2017:run 3 | train error = 0.0049 | valid error = 0.4829 | test error = 0.4718
Sun Mar 19 01:46:17 2017:run 4 | train error = 0.0316 | valid error = 0.5111 | test error = 0.4708
Sun Mar 19 05:01:34 2017:run 5 | train error = 0.0235 | valid error = 0.5101 | test error = 0.4970
Sun Mar 19 08:16:15 2017:run 6 | train error = 0.0065 | valid error = 0.4103 | test error = 0.4294
Sun Mar 19 11:30:52 2017:run 7 | train error = 0.0177 | valid error = 0.4486 | test error = 0.4335
Sun Mar 19 14:36:37 2017:run 8 | train error = 0.0000 | valid error = 0.0010 | test error = 0.0010
Sun Mar 19 17:46:36 2017:run 9 | train error = 0.0000 | valid error = 0.0010 | test error = 0.0000
Sun Mar 19 21:01:10 2017:run 10 | train error = 0.0000 | valid error = 0.0030 | test error = 0.0020
Sun Mar 19 21:01:12 2017:final test error = 0.001008064516129
```
As we can see 4/10 runs solved the task and 6/10 overfitted.

Another snippet:
```
$ cat task\=16-seed\=3912-model\=entnet-embed\=icmul+bow-edim\=100-initstd\=0.1-sdt_decay\=25-optim\=adam-sdt\=0.01-nhop\=1-tied\=0-T\=70-memslots\=20.log 
Mon Mar 20 15:03:53 2017:run 1 | train error = 0.0004 | valid error = 0.4415 | test error = 0.4224
Mon Mar 20 18:48:54 2017:run 2 | train error = 0.0202 | valid error = 0.5060 | test error = 0.4879
Mon Mar 20 22:07:02 2017:run 3 | train error = 0.0285 | valid error = 0.5363 | test error = 0.4970
Tue Mar 21 01:22:41 2017:run 4 | train error = 0.0171 | valid error = 0.4385 | test error = 0.4224
Tue Mar 21 04:38:20 2017:run 5 | train error = 0.0054 | valid error = 0.4758 | test error = 0.4657
Tue Mar 21 07:54:01 2017:run 6 | train error = 0.0102 | valid error = 0.5242 | test error = 0.4819
Tue Mar 21 11:09:42 2017:run 7 | train error = 0.0212 | valid error = 0.4738 | test error = 0.4708
Tue Mar 21 14:54:28 2017:run 8 | train error = 0.0188 | valid error = 0.5373 | test error = 0.5050
Tue Mar 21 18:57:37 2017:run 9 | train error = 0.0088 | valid error = 0.5000 | test error = 0.4778
Tue Mar 21 22:13:33 2017:run 10 | train error = 0.0229 | valid error = 0.4496 | test error = 0.4597
Tue Mar 21 22:13:36 2017:final test error = 0.42237903225806
```

PS: The exact code to run this is here: https://github.com/botev/MemNN. I've needed to modify their script as they had a hardcoded `seed=123` for parameter intialization and the arguments to the script were not affecting it. The only other difference is I'm also printing the shapes of all parameters (This helped we understand exactly their model and reproduce it in Theano). You can just check the single commit I did that it does not change anything else at all.
